crypt.go

```diff
@@ -0,0 +1,488 @@
+	return fmt.Sprintf("%s failed: %s", e.path, e.err)
+}
+
+func (e *execError) Unwrap() error {
+	return e.err
+}
+
+func wrapExecError(cmd *exec.Cmd, err error) error {
+	if err == nil {
+		return nil
+	}
+	return &execError{path: cmd.Path, err: err}
+}
+
+func isExecError(err error, path string) bool {
+	var e *execError
+	return xerrors.As(err, &e) && e.path == path
+}
+
+func activate(volumeName, sourceDevicePath string, key []byte, options []string) error {
+	keyFilePath, err := func() (string, error) {
+		f, err := ioutil.TempFile(runDir, filepath.Base(os.Args[0])+".")
+		if err != nil {
+			return "", xerrors.Errorf("cannot create temporary file: %w", err)
+		}
+		defer f.Close()
+		if err := f.Chmod(0600); err != nil {
+			return "", err
+		}
+		if _, err := f.Write(key); err != nil {
+			return "", xerrors.Errorf("cannot write key to file: %w", err)
+		}
+		return f.Name(), nil
+	}()
+	if err != nil {
+		return xerrors.Errorf("error saving key for systemd-cryptsetup: %w", err)
+	}
+	defer os.Remove(keyFilePath)
+
+	cmd := exec.Command(systemdCryptsetupPath, "attach", volumeName, sourceDevicePath, keyFilePath, strings.Join(options, ","))
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "SYSTEMD_LOG_TARGET=console")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return xerrors.Errorf("cannot create stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return xerrors.Errorf("cannot create stderr pipe: %w", err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	done := make(chan bool, 2)
+	go func() {
+		rd := bufio.NewScanner(stdout)
+		for rd.Scan() {
+			fmt.Printf("systemd-cryptsetup: %s\n", rd.Text())
+		}
+		done <- true
+	}()
+	go func() {
+		rd := bufio.NewScanner(stderr)
+		for rd.Scan() {
+			fmt.Fprintf(os.Stderr, "systemd-cryptsetup: %s\n", rd.Text())
+		}
+		done <- true
+	}()
+	for i := 0; i < 2; i++ {
+		<-done
+	}
+
+	return wrapExecError(cmd, cmd.Wait())
+}
+
+func askPassword(sourceDevicePath, msg string) (string, error) {
+	cmd := exec.Command(
+		"systemd-ask-password",
+		"--icon", "drive-harddisk",
+		"--id", filepath.Base(os.Args[0])+":"+sourceDevicePath,
+		msg)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	cmd.Stdin = os.Stdin
+	if err := cmd.Run(); err != nil {
+		return "", wrapExecError(cmd, err)
+	}
+	result, err := out.ReadString('\n')
+	if err != nil {
+		return "", xerrors.Errorf("cannot read result from systemd-ask-password: %w", err)
+	}
+	return strings.TrimRight(result, "\n"), nil
+}
+
+func getPassword(sourceDevicePath, description string, reader io.Reader) (string, error) {
+	if reader != nil {
+		scanner := bufio.NewScanner(reader)
+		switch {
+		case scanner.Scan():
+			return scanner.Text(), nil
+		case scanner.Err() != nil:
+			return "", xerrors.Errorf("cannot obtain %s from scanner: %w", description, scanner.Err())
+		}
+	}
+	return askPassword(sourceDevicePath, "Please enter the "+description+" for disk "+sourceDevicePath+":")
```

Comment:
the *Options should allow a way to override this function, something like:

```
AskPassword func(kind PasswordKind, sourceDevicePath string) (string, error)
```

where `PasswordKind`:

```
type PasswordKind int

const (
  PasswordKindPIN =  iota
  PasswordKindRecoveryKey
)
```

can be a follow up though